File: source/aoBlogs2/Views/ArchiveView.vb

Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Text.RegularExpressions
Imports Contensive.Addons.Blog.Controllers
Imports Contensive.Addons.Blog.Models
Imports Contensive.BaseClasses

Namespace Views
    Public Class ArchiveView
        '
        '
        '====================================================================================
        '
        Public Shared Function GetFormBlogArchiveDateList(cp As CPBaseClass, app As ApplicationEnvironmentModel, request As View.RequestModel) As String
            '
            Dim result As String = ""
            Try

                Dim blog As BlogModel = app.blog
                Dim blogEntry As BlogPostModel = app.blogEntry
                Dim OpenSQL As String = ""
                Dim contentControlId As String = cp.Content.GetID(cnBlogEntries).ToString
                '
                result = vbCrLf & cp.Content.GetCopy("Blogs Archives Header for " & Blog.name, "<h2>" & Blog.name & " Blog Archive</h2>")
                ' 
                Dim archiveDateList As List(Of BlogCopyModel.ArchiveDateModel) = BlogCopyModel.createArchiveListFromBlogCopy(cp, Blog.Id)
                If (archiveDateList.Count = 0) Then
                    '
                    ' No archives, give them an error
                    result &= "<div class=""aoBlogProblem"">There are no current blog entries</div>"
                    result &= "<div class=""aoBlogFooterLink""><a href=""" & app.blogPageBaseLink & """>" & BackToRecentPostsMsg & "</a></div>"
                Else
                    Dim ArchiveMonth As Integer
                    Dim ArchiveYear As Integer
                    If archiveDateList.Count = 1 Then
                        '
                        ' one archive - just display it
                        ArchiveMonth = archiveDateList.First.Month
                        ArchiveYear = archiveDateList.First.Year
                        result &= GetFormBlogArchivedBlogs(cp, app, request)
                    Else
                        '
                        ' Display List of archive
                        Dim qs As String = cp.Utils.ModifyQueryString(app.blogListLink, RequestNameSourceFormID, FormBlogArchiveDateList.ToString())
                        qs = cp.Utils.ModifyQueryString(qs, RequestNameFormID, FormBlogArchivedBlogs.ToString())
                        For Each archiveDate In archiveDateList
                            ArchiveMonth = archiveDate.Month
                            ArchiveYear = archiveDate.Year
                            Dim NameOfMonth As String = MonthName(ArchiveMonth)
                            qs = cp.Utils.ModifyQueryString(qs, RequestNameArchiveMonth, CStr(ArchiveMonth))
                            qs = cp.Utils.ModifyQueryString(qs, RequestNameArchiveYear, CStr(ArchiveYear))
                            result &= vbCrLf & vbTab & vbTab & "<div class=""aoBlogArchiveLink""><a href=""?" & qs & """>" & NameOfMonth & " " & ArchiveYear & "</a></div>"
                        Next
                        result &= vbCrLf & vbTab & "<div class=""aoBlogFooterLink""><a href=""" & app.blogPageBaseLink & """>" & BackToRecentPostsMsg & "</a></div>"
                    End If
                End If
                '

            Catch ex As Exception
                cp.Site.ErrorReport(ex)
            End Try
            Return result
        End Function
        '
        '====================================================================================
        '
        Public Shared Function GetFormBlogArchivedBlogs(cp As CPBaseClass, app As ApplicationEnvironmentModel, request As View.RequestModel) As String
            '
            Dim result As String = ""
            Try
                Dim blog As BlogModel = app.blog
                Dim blogEntry As BlogPostModel = app.blogEntry
                Dim PageNumber As Integer
                '
                ' If it is the current month, start at entry 6
                '
                PageNumber = 1
                '
                ' List Blog Entries
                '
                Dim BlogEntryModelList As List(Of BlogPostModel) = DbModel.createList(Of BlogPostModel)(cp, "(Month(DateAdded) = " & request.ArchiveMonth & ")And(year(DateAdded)=" & request.ArchiveYear & ")And(BlogID=" & Blog.Id & ")", "DateAdded Desc")
                If (BlogEntryModelList.Count = 0) Then
                    result = "<div Class=""aoBlogProblem"">There are no blog archives For " & request.ArchiveMonth & "/" & request.ArchiveYear & "</div>"
                Else
                    Dim EntryPtr As Integer = 0
                    Dim entryEditLink As String = ""
                    For Each blogEntry In BlogEntryModelList
                        Dim EntryID As Integer = blogEntry.id
                        Dim AuthorMemberID As Integer = blogEntry.AuthorMemberID
                        If AuthorMemberID = 0 Then
                            AuthorMemberID = blogEntry.CreatedBy
                        End If
                        Dim DateAdded As Date = blogEntry.DateAdded
                        Dim EntryName As String = blogEntry.name
                        If app.userIsEditing Then
                            entryEditLink = cp.Content.GetEditLink(EntryName, EntryID.ToString(), True, EntryName, True)
                        End If
                        Dim EntryCopy As String = blogEntry.copy
                        Dim allowComments As Boolean = blogEntry.AllowComments
                        Dim BlogTagList As String = blogEntry.tagList
                        Dim primaryImagePositionId As Integer = blogEntry.primaryImagePositionId
                        Dim Return_CommentCnt As Integer
                        result &= BlogEntryCellView.getBlogPostCell(cp, app, blogEntry, False, False, Return_CommentCnt, BlogTagList)
                    Next
                    result &= "<hr>"
                    EntryPtr = EntryPtr + 1

                End If
                '              
                '
                Dim qs As String = app.blogListLink
                'qs = cp.Utils.ModifyQueryString(qs, RequestNameFormID, FormBlogSearch.ToString())
                result &= "<div>&nbsp;</div>"
                'result &= "<div Class=""aoBlogFooterLink""><a href=""?" & qs & """>Search</a></div>"
                qs = cp.Doc.RefreshQueryString()
                qs = cp.Utils.ModifyQueryString(qs, RequestNameBlogEntryID, "", True)
                qs = cp.Utils.ModifyQueryString(qs, RequestNameFormID, FormBlogPostList.ToString())
                result &= "<div Class=""aoBlogFooterLink""><a href=""" & app.blogPageBaseLink & """>" & BackToRecentPostsMsg & "</a></div>"
                Call cp.CSNew.Close()
                result &= cp.Html.Hidden(RequestNameSourceFormID, FormBlogArchivedBlogs.ToString())
                result = cp.Html.Form(result)
                '
                GetFormBlogArchivedBlogs = result
                '
            Catch ex As Exception
                cp.Site.ErrorReport(ex)
            End Try
            Return result
        End Function
        '
    End Class
End Namespace

File: source/aoBlogs2/Views/ArchiveView.cs

using Contensive.BaseClasses;
using Contensive.Blog.Controllers;
using Contensive.Blog.Models;
using Contensive.Models.Db;
using Microsoft.VisualBasic;
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;

namespace Contensive.Blog.Views {
    public class ArchiveView {
        // 
        // 
        // ====================================================================================
        // 
        public static string getFormBlogArchiveDateList(CPBaseClass cp, ApplicationEnvironmentModel app, Models.View.RequestModel request) {
            // 
            string result = "";
            try {

                var blog = app.blog;
                var blogEntry = app.blogPost;
                string contentControlId = cp.Content.GetID(constants.cnBlogEntries).ToString();
                // 
                result = Constants.vbCrLf + cp.Content.GetCopy("Blogs Archives Header for " + blog.name, "<h1>" + blog.name + " Archives</h1>");
                // 
                var archiveDateList = BlogEntryModel.createArchiveListFromBlogCopy(cp, blog.id);
                if (archiveDateList.Count == 0) {
                    // 
                    // No archives, give them an error
                    result += "<div class=\"aoBlogProblem\">There are no current blog entries</div>";
                    result += "<div class=\"aoBlogFooterLink\"><a href=\"" + app.blogPageBaseLink + "\">" + constants.BackToRecentPostsMsg + "</a></div>";
                } else {
                    int ArchiveMonth;
                    int ArchiveYear;
                    if (archiveDateList.Count == 1) {
                        // 
                        // one archive - just display it
                        ArchiveMonth = archiveDateList.First().Month;
                        ArchiveYear = archiveDateList.First().Year;
                        result += getFormBlogArchivedBlogs(cp, app, request);
                    } else {
                        // 
                        // Display List of archive
                        string qs = cp.Utils.ModifyQueryString(app.blogBaseLink, constants.RequestNameSourceFormID, constants.FormBlogArchiveDateList.ToString());
                        qs = cp.Utils.ModifyQueryString(qs, constants.rnFormID, constants.FormBlogArchivedBlogs.ToString());
                        foreach (var archiveDate in archiveDateList) {
                            ArchiveMonth = archiveDate.Month;
                            ArchiveYear = archiveDate.Year;
                            string NameOfMonth = DateAndTime.MonthName(ArchiveMonth);
                            qs = cp.Utils.ModifyQueryString(qs, constants.RequestNameArchiveMonth, ArchiveMonth.ToString());
                            qs = cp.Utils.ModifyQueryString(qs, constants.RequestNameArchiveYear, ArchiveYear.ToString());
                            result += Constants.vbCrLf + Constants.vbTab + Constants.vbTab + "<div class=\"aoBlogArchiveLink\"><a href=\"" + qs + "\">" + NameOfMonth + " " + ArchiveYear + "</a></div>";
                        }
                        result += Constants.vbCrLf + Constants.vbTab + "<div class=\"aoBlogFooterLink\"><a href=\"" + app.blogPageBaseLink + "\">" + constants.BackToRecentPostsMsg + "</a></div>";
                    }
                }
            }
            // 

            catch (Exception ex) {
                cp.Site.ErrorReport(ex);
            }
            return result;
        }
        // 
        // ====================================================================================
        // 
        public static string getFormBlogArchivedBlogs(CPBaseClass cp, ApplicationEnvironmentModel app, Models.View.RequestModel request) {
            string getFormBlogArchivedBlogsRet;
            //
            string result = "";
            try {
                var blog = app.blog;
                // 
                // List Blog Entries
                // 
                cp.Doc.AddMetaDescription($"Archives {DateTimeFormatInfo.CurrentInfo.GetMonthName(request.ArchiveMonth)} {request.ArchiveYear}, {MetadataController.getBlogMetaDescription(app, blog)}");
                string title = $"Archives {DateTimeFormatInfo.CurrentInfo.GetMonthName(request.ArchiveMonth)} {request.ArchiveYear}, {MetadataController.getBlogMetaTitle(app, blog)} ";
                MetadataController.addTitle(cp, title);
                //
                result += $"<h1>{title}</h1>";
                var postList = DbBaseModel.createList<BlogEntryModel>(cp, "(Month(COALESCE(datePublished, dateAdded)) = " + request.ArchiveMonth + ")And(year(COALESCE(datePublished, dateAdded))=" + request.ArchiveYear + ")And(BlogID=" + blog.id + ")", "COALESCE(datePublished, dateAdded) Desc");
                if (postList.Count == 0) {
                    // 
                    // -- no results
                    result += $"<div class=\"aoBlogProblem\">There are no blog archives For {request.ArchiveMonth}/{request.ArchiveYear}</div>";
                    //
                    // -- in this case, there may have been a link alias that led here, remove it
                    string linkAliasQs = $"FormID=600&ArchiveMonth={request.ArchiveMonth}&ArchiveYear={request.ArchiveYear}";
                    DbBaseModel.deleteRows<LinkAliasModel>(cp, $"(pageid={cp.Doc.PageId}) and (queryStringSuffix={cp.Db.EncodeSQLText(linkAliasQs)})");
                } else {
                    // 
                    // -- list of articles
                    int EntryPtr = 0;
                    string entryEditLink = "";
                    var Return_CommentCnt = default(int);
                    foreach (var blogPost in postList) {
                        int EntryID = blogPost.id;
                        int AuthorMemberID = blogPost.authorMemberId;
                        if (AuthorMemberID == 0) {
                            AuthorMemberID = cp.Utils.EncodeInteger(blogPost.createdBy);
                        }
                        var dateAdded = blogPost.dateAdded;
                        string EntryName = blogPost.name;
                        if (app.userIsEditing) {
                            entryEditLink = cp.Content.GetEditLink(EntryName, EntryID.ToString(), true, EntryName, true);
                        }
                        string EntryCopy = blogPost.copy;
                        bool allowComments = blogPost.allowComments;
                        string BlogTagList = blogPost.tagList;
                        int primaryImagePositionId = blogPost.primaryImagePositionId;
                        List<BlogImageModel> blogImageList = BlogImageModel.getPostImageList(cp, blogPost);
                        result += BlogEntryCellView.getBlogPostCell(cp, app, blogPost, blogImageList, false, false, Return_CommentCnt, entryEditLink);
                        result += "<hr>";
                    }
                    EntryPtr++;

                }
                // 
                // 
                string qs = app.blogBaseLink;
                // qs = cp.Utils.ModifyQueryString(qs, RequestNameFormID, FormBlogSearch.ToString())
                result += "<div>&nbsp;</div>";
                // result &= "<div Class=""aoBlogFooterLink""><a href=""?" & qs & """>Search</a></div>"
                qs = cp.Doc.RefreshQueryString;
                qs = cp.Utils.ModifyQueryString(qs, constants.RequestNameBlogEntryID, "", true);
                qs = cp.Utils.ModifyQueryString(qs, constants.rnFormID, constants.FormBlogPostList.ToString());
                result += "<div Class=\"aoBlogFooterLink\"><a href=\"" + app.blogPageBaseLink + "\">" + constants.BackToRecentPostsMsg + "</a></div>";
                result += cp.Html.Hidden(constants.RequestNameSourceFormID, constants.FormBlogArchivedBlogs.ToString());
                result = cp.Html.Form(result);
                // 
                getFormBlogArchivedBlogsRet = result;
            }
            // 
            catch (Exception ex) {
                cp.Site.ErrorReport(ex);
            }
            return result;
        }
        // 
    }
}